### PR TITLE
Docs/autogenerated typedocs

### LINKED ISSUE #1362: docs(dmt-core): add initial docs setup to dmt-core
Some possible alternatives to look at:

- remix (react router)
- doctave (template)
- gatsbyjs (reactjs + react-hook-form)
- docz
- storybook
- docusaurus
and more...

Also see #1358

### LINKED ISSUE #1363: docs(dmt-core): generate docs automatically for dm-core

### PR DESCRIPTION
**Pull request by netr0m**
## What does this pull request change?
- Initial support for automatically generated MDX documentation for
  - Components
  - Hooks
  - Enums
  - Types

## Issues related to this change
- Resolves #1363 
- Resolves #1362

### CODE REVIEW

**Comment by netr0m** [comment]
Component Example:


Enum Example:


Hook Example:


Type Example:

**Comment by netr0m** [comment]
For later:
- Should probably parse typedocs JSON => convert into smaller JSON containing only the (processed) data that we need, in order to
  - Avoid huge JSON files in source (currently ~69k lines)
  - Avoid having to parse the huge JSON for each render

- Content:
  - Table:
    - Display different fields based on type, e.g.
      - Enum:
        - Key / Name, Value, Description
      - Component / Hook / Type:
        - Name, Type, Optional, Description
  - Show "Returns" only for functions, hooks, components
  - Parameters
    - Use "Members" for Enums
    - Use "Properties" for Types?
    - Keep "Parameters" for functions, hooks, components
- Configure pre-commit / linting (prettier, etc) to ensure style

**Comment by KristianKjerstad** [comment]
small bug: duplicate Home

image

**Comment by KristianKjerstad** [comment]
not sure if its possible, but instead of prepending with props, props.children, props.closeScrim, etc, it would be more clean to just use the prop names: children, closeScrim, etc.

Also, the headline "Props" might be more suited than "Parameters"

image

**Comment by KristianKjerstad** [comment]
is it possible to rename Parameters to Attributes or Properties? (Not 100% what the correct typescript term for these are) 

image

File: web/packages/dmt-core/src/utils/truncatePathString.ts [hunk 1]
```diff
@@ -17,10 +17,10 @@ export function truncatePathString(
...
   while (truncatedPath.length <= maxLength) {
     previousPath = truncatedPath
-    truncatedPath = truncatedPath.replace(
-      '...',
-      `${splitTextOnFolders[counter]}/...`
-    )
+    //truncatedPath = truncatedPath.replace(
+    //  '...',
+    //  `${splitTextOnFolders[counter]}/...`
+    // )
```
Comment (KristianKjerstad) [inline web/packages/dmt-core/src/utils/truncatePathString.ts:23]: is there a bug here? maybe create new issue for it

**Review by KristianKjerstad — APPROVED** [review]
Looks good!

File: web/packages/dmt-core/src/utils/truncatePathString.ts (hunk 1 above)
Comment (netr0m) [reply web/packages/dmt-core/src/utils/truncatePathString.ts:23]: IDK :man_shrugging:  Came from the rebase :incoming_envelope:

**Review by soofstad — CHANGES_REQUESTED** [review]
If possible; avoid commiting typedocs.json

File: web/packages/dmt-core/src/components/Pickers/BlueprintPicker.tsx [hunk 2]
```diff
@@ -14,13 +14,38 @@ import {
...
+  /** Whether the input should be disabled */
   disabled?: boolean
+  /** A title for the picker */
   label?: string
-}) => {
+}
+
+/**
+ * Component which renders a blueprint picker,
+ * allowing the user to select a blueprint
```
Comment (soofstad) [inline web/packages/dmt-core/src/components/Pickers/BlueprintPicker.tsx:32]:
```suggestion
 * allowing the user to select a reference to a blueprint
```

File: web/packages/dmt-core/src/hooks/useBlueprint.tsx [hunk 3]
```diff
@@ -2,6 +2,28 @@ import { useContext, useEffect, useState } from 'react'
...
+ * const [blueprint, isLoading, error] = useBlueprint('someType')
+ * if (isLoading) return <div>Loading...</div>
+ * if (error) return <div>Error getting the blueprint</div>
+ *
+ * if (blueprint) {
+ *   return (<p>Blueprint: {blueprint?.name} ({blueprint?.type})</p>)
+ * }
+ * ```
+ *
+ * @param typeRef - The type of blueprint to retrieve
```
Comment (soofstad) [inline web/packages/dmt-core/src/hooks/useBlueprint.tsx:24]:
```suggestion
 * @param typeRef - The reference to the blueprint to retrieve
```

File: web/packages/dmt-core/src/hooks/useBlueprint.tsx [hunk 4]
```diff
@@ -2,6 +2,28 @@ import { useContext, useEffect, useState } from 'react'
...
+ * @returns A list containing the blueprint document, a boolean representing the state, and an Error, if any.
```
Comment (soofstad) [inline web/packages/dmt-core/src/hooks/useBlueprint.tsx:25]:
```suggestion
 * @returns A list containing the blueprint document, a boolean representing the loading state, and an Error, if any.
```

File: web/packages/dmt-core/src/hooks/useDocument.tsx [hunk 5]
```diff
@@ -6,12 +6,36 @@ import { NotificationManager } from 'react-notifications'
...
+ *   console.error(error)
+ *   return <div>Error getting the document</div>
+ * }
+ *
+ * <DisplayDocument document={document} />
+ * ```
+ *
+ * @param dataSourceId The ID of the data source
+ * @param documentId The ID of the document
+ * @param depth The maximum recursion depth
```
Comment (soofstad) [inline web/packages/dmt-core/src/hooks/useDocument.tsx:37]:
```suggestion
 * @param depth The maximum depth level of nested objects to resolve
```

File: web/packages/dmt-core/src/hooks/useDocument.tsx [hunk 6]
```diff
@@ -6,12 +6,36 @@ import { NotificationManager } from 'react-notifications'
...
+ * @returns A list containing the document, a boolean representing the state, a function to update the document, and an Error, if any.
```
Comment (soofstad) [inline web/packages/dmt-core/src/hooks/useDocument.tsx:38]:
```suggestion
 * @returns A list containing the document, a boolean representing the loading state, a function to update the document, and an Error, if any.
```

**Review by soofstad — CHANGES_REQUESTED** [review]

**Review by soofstad — APPROVED** [review]

**Review by soofstad — APPROVED** [review]

**Pull request merged**